### PR TITLE
refactor(completion): improve popup window position 

### LINKED ISSUE #27768: completeopt=popup crashing neovim
### Problem

When using

```lua
vim.o.completeopt = 'menuone,noinsert,popup'
```

vim is sometimes crashing on either free or malloc issues. This issue is not very frequent but it never happens without popup enabled. I have no issues or crashes with any other floating windows. With popup enabled it looks like it is most easily reproducable when completing vim.keymap.set (but still do not happens that often, but after minute or two trying it will).

image

### Steps to reproduce

How I got this issue:
- Use this plugin - https://github.com/deathbeam/autocomplete.nvim/blob/main/lua/autocomplete/buffer.lua
- Configure:
```
vim.o.completeopt = 'menuone,noinsert,popup'
require('autocomplete.buffer').setup()
```
- Start typing using Lua lsp
```
vim.keymap.set
```
- Repeat/delete chars until bug occcurs

What the plugin does is when completion changes it:
1. Does completionItem/resolve
2. Closes current popup if open (different bug but it do not updates its size,but I saw the bug happen wthout this as well)
3. Calls
```
local wininfo = vim.api.nvim_complete_set(selected, { info = value })
if wininfo.winid and wininfo.bufnr then
    vim.wo[wininfo.winid].conceallevel = 2
    vim.wo[wininfo.winid].concealcursor = 'niv'
    vim.bo[wininfo.bufnr].syntax = 'markdown'
    return true
end
```

I suspect that nvim_complete_set is causing the issue. Im 90% sure the plugin is implemented correctly but even if it isnt opening popup window should not be crashing neovim.

### Expected behavior

No crashing

### Neovim version (nvim -v)

NVIM v0.10.0-dev-2531+gdc2379b89

### Vim (not Nvim) behaves the same?

no

### Operating system/version

Arch Linux x86_64

### Terminal name/version

st

### $TERM environment variable

tmux-256color

### Installation

repo

### PR DESCRIPTION
**Pull request by glepnir**
Problem: 
1. when new leader add and selected is -1 the pum window width changed like (be small) the info floating window position not changed.
2.  correct selected check avoid memory leak.
3. The info string does not need to be stored in compl, it is usually a very long byte string. So a large amount of memory needs to be allocated. We just need to write them into the info buffer.

Solutioin: when new leader add and selected is -1 we still need adjust info window width. And don't store info value. 

Fix https://github.com/neovim/neovim/issues/27768

### CODE REVIEW

**Comment by zeertzjq** [comment]
lint failure

File: src/nvim/winfloat.c [hunk 1]
```diff
@@ -337,3 +339,49 @@ win_T *win_float_find_altwin(const win_T *win, const tabpage_T *tp)
...
+  config.style = kWinStyleMinimal;
+  Error err = ERROR_INIT;
+  win_T *wp = win_new_float(NULL, true, config, &err);
+  if (!wp) {
+    return NULL;
+  }
+  if (new_buf) {
+    Buffer b = nvim_create_buf(false, true, &err);
+    if (!b) {
+      win_free(curwin, NULL);
```
Comment (zeertzjq) [inline src/nvim/winfloat.c:367]:
```suggestion
      win_free(wp, NULL);
```

Comment (zeertzjq) [reply src/nvim/winfloat.c:367]: And `win_remove()` should be needed before this.

**Comment by glepnir** [comment]
can someone review this ?

**Comment by aorith** [comment]
Hello,

I've tested the PR, with master neovim it crashes easily following the steps of the issue:

```
NVIM v0.10.0-dev-2961+g4d52b0cf6
Build type: RelWithDebInfo
LuaJIT 2.1.1713484068
```

With this PR it doesn't:

```
NVIM v0.10.0-dev-2962+gca3fbcbce
Build type: RelWithDebInfo
LuaJIT 2.1.1713484068
```

File: src/nvim/winfloat.c [hunk 2]
```diff
@@ -337,3 +340,52 @@ win_T *win_float_find_altwin(const win_T *win, const tabpage_T *tp)
...
+  config.relative = kFloatRelativeEditor;
+  config.focusable = false;
+  config.anchor = 0;  // NW
+  config.noautocmd = true;
+  config.hide = true;
+  config.style = kWinStyleMinimal;
+  Error err = ERROR_INIT;
+
+  block_autocmds();
+  win_T *wp = win_new_float(NULL, true, config, &err);
```
Comment (zeertzjq) [inline src/nvim/winfloat.c:364]:
```suggestion
  win_T *wp = win_new_float(NULL, false, config, &err);
```

File: src/nvim/winfloat.c [hunk 3]
```diff
@@ -337,3 +340,52 @@ win_T *win_float_find_altwin(const win_T *win, const tabpage_T *tp)
...
+  config.anchor = 0;  // NW
+  config.noautocmd = true;
+  config.hide = true;
+  config.style = kWinStyleMinimal;
+  Error err = ERROR_INIT;
+
+  block_autocmds();
+  win_T *wp = win_new_float(NULL, false, config, &err);
+  if (!wp) {
+    block_autocmds();
```
Comment (zeertzjq) [inline src/nvim/winfloat.c:366]:
```suggestion
    unblock_autocmds();
```

File: src/nvim/winfloat.c [hunk 4]
```diff
@@ -337,3 +340,52 @@ win_T *win_float_find_altwin(const win_T *win, const tabpage_T *tp)
...
+    return NULL;
+  }
+
+  if (new_buf) {
+    Buffer b = nvim_create_buf(false, true, &err);
+    if (!b) {
+      win_remove(wp, NULL);
+      win_free(wp, NULL);
+      block_autocmds();
```
Comment (zeertzjq) [inline src/nvim/winfloat.c:375]:
```suggestion
      unblock_autocmds();
```

Comment (glepnir) [reply src/nvim/winfloat.c]: emm sry for this two....

**Pull request merged**